### PR TITLE
Support ListAgg WITHIN GROUP clause

### PR DESCRIPTION
**Pull request by xiangfu0**
- Support `WITHIN GROUP` Clause in MultiStage Engine
  - Use `PinotLogicalSortExchange` for WITHIN GROUP instead of `PinotLogicalExchange` for leaf data exchange
- Support `ListAgg` aggregation function
- Support passing `Distinct` flag to `ListAgg` implementation

Limitation:
- Only one `WITHIN GROUP` Clause is supported right now.

Sample query:
```
SELECT 
    Origin, Dest,
    LISTAGG(distinct Carrier, ', ') WITHIN GROUP (ORDER BY Carrier)
FROM 
    airlineStats
GROUP BY 
    Dest, Origin
```
Query Plan:
```
Execution Plan
LogicalProject(Origin=[$1], Dest=[$0], EXPR$2=[$2])
  LogicalAggregate(group=[{0, 1}], EXPR$2=[LISTAGG(DISTINCT $2, $3) WITHIN GROUP ([2])])
    PinotLogicalSortExchange(distribution=[hash[0, 1]], collation=[[2]], isSortOnSender=[false], isSortOnReceiver=[true])
      LogicalProject(Dest=[$28], Origin=[$61], Carrier=[$17], $f3=[_UTF-8', '])
        LogicalTableScan(table=[[default, airlineStats]])
```

image

```
SELECT
    Carrier,
    LISTAGG(DISTINCT CONCAT(Origin, Dest, ' | '), ', ') WITHIN GROUP (ORDER BY Origin, Dest)
FROM airlineStats
GROUP BY Carrier
ORDER BY Carrier
```

Query Plan
```
Execution Plan
LogicalSort(sort0=[$0], dir0=[ASC])
  PinotLogicalSortExchange(distribution=[hash], collation=[[0]], isSortOnSender=[false], isSortOnReceiver=[true])
    LogicalAggregate(group=[{0}], EXPR$1=[LISTAGG(DISTINCT $1, $2) WITHIN GROUP ([3, 4])])
      PinotLogicalSortExchange(distribution=[hash[0]], collation=[[3, 4]], isSortOnSender=[false], isSortOnReceiver=[true])
        LogicalProject(Carrier=[$17], $f1=[CONCAT($61, $28, _UTF-8' | ')], $f2=[_UTF-8', '], Origin=[$61], Dest=[$28])
          LogicalTableScan(table=[[default, airlineStats]])
```
image

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/FunctionContext.java [hunk 1]
```diff
@@ -35,12 +35,14 @@ public enum Type {
   private final Type _type;
   private String _functionName;
   private final List<ExpressionContext> _arguments;
+  private final boolean _isDistinct;
 
-  public FunctionContext(Type type, String functionName, List<ExpressionContext> arguments) {
+  public FunctionContext(Type type, String functionName, List<ExpressionContext> arguments, boolean isDistinct) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/request/context/FunctionContext.java:40]: This is single-stage engine context. Do we need to push this to the single stage side? I think we will need to collect all records for a partition, sorting the records, then apply the distinct and aggregation?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.function.BaseSingleInputAggregationFunction;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public class ListAggFunction
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:36]: Do we need to implement it as a single-stage aggregation? I think it should follow the way how window function is implemented

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/plannode/AggregateNode.java [hunk 3]
```diff
@@ -42,6 +45,12 @@ public class AggregateNode extends AbstractPlanNode {
   private List<RexExpression> _groupSet;
   @ProtoProperties
   private AggType _aggType;
+  @ProtoProperties
+  private List<RexExpression> _collationKey = ImmutableList.of();
+  @ProtoProperties
+  private List<RelFieldCollation.Direction> _collationDirection = ImmutableList.of();
+  @ProtoProperties
+  private List<RelFieldCollation.NullDirection> _collationNullDirection = ImmutableList.of();
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/plannode/AggregateNode.java:53]: Do you put them as 3 separate list because we cannot serialize `RelFieldCollation` properly? Some comments will help

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        return;
+      }
+    }
+    aggregateArray(length, valueSet, values);
+  }
+
+  private void aggregateArray(int length, ObjectArrayList<String> valueSet, String[] values) {
+    for (int i = 0; i < length; i++) {
+      String value = values[i];
+      if (_isDistinct && valueSet.contains(value)) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:89]: This is `O(n)` contains, which will be extremely slow

**Review by Jackie-Jiang — COMMENTED** [review]
We should generalize the query flow for aggregate with `distinct` or `within group order by`. It should be modeled as projection/distinct on the leaf stage instead of aggregation. The aggregation should only be applied on the intermediate stage, which is similar to the window function handling logic

**Comment by xiangfu0** [comment]
> We should generalize the query flow for aggregate with `distinct` or `within group order by`. It should be modeled as projection/distinct on the leaf stage instead of aggregation. The aggregation should only be applied on the intermediate stage, which is similar to the window function handling logic

Do you mean We should have a new rule or reuse existing rule to modify the query plan?

**Comment by xiangfu0** [comment]
> We should generalize the query flow for aggregate with `distinct` or `within group order by`. It should be modeled as projection/distinct on the leaf stage instead of aggregation. The aggregation should only be applied on the intermediate stage, which is similar to the window function handling logic

I think the distinct will still happen in both leaf and intermediate stage, as there could still be duplicates across leaf nodes.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java]: there is no good way for this, unless we keep two copies, one is the array list for ordering, the other is a hash map for collision checks

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:37]: Right now I model it as AggType.DIRECT, so LEAF stage agg will be used in intermediate stage.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/plannode/AggregateNode.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/plannode/AggregateNode.java:54]: done

File: pinot-common/src/main/java/org/apache/pinot/common/request/context/FunctionContext.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/request/context/FunctionContext.java:40]: this is also used in multi-stage engine to init the function.

**Comment by xiangfu0** [comment]
Simplified the code to for both collation and distinct handling.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        String value = values[i];
+        if (_isDistinct) {
+          if (distinctValueSet.contains(value)) {
+            continue;
+          } else {
+            distinctValueSet.add(value);
+          }
+        }
+        valueSet.add(value);
+      }
```
Comment (gortiz) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:116]: Can you use the utility functions included in `NullableSingleInputAggregationFunction`? They have been proven to be faster than this direct approach when we check the bitmap on each iteration.

In this case you will probably need to use `foldNotNull`

Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java]: done, used `forEachNotNull`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    ObjectArrayList<String> valueSet = getObjectArrayList(aggregationResultHolder);
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      ObjectOpenHashSet<String> distinctValueSet = new ObjectOpenHashSet<>();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:71]: This is per block set. The values from previous blocks won't be carried over

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  @Override
+  public void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      for (int i = from; i < to; i++) {
+        for (int groupKey : groupKeysArray[i]) {
+          ObjectArrayList<String> groupValueList = getObjectArrayList(groupByResultHolder, groupKey);
+          if (_isDistinct && groupValueList.contains(values[i])) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:130]: This is still O(n) check

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  @Override
+  public void aggregateGroupBySV(int length, int[] groupKeyArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      for (int i = from; i < to; i++) {
+        String value = values[i];
+        ObjectArrayList<String> groupValueList = getObjectArrayList(groupByResultHolder, groupKeyArray[i]);
+        if (_isDistinct && groupValueList.contains(value)) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:113]: This is still O(n) check

**Review by Jackie-Jiang — COMMENTED** [review]
The difference for `ListAggFunction` and `ArrayAggString`/`ArrayAggDistinctStringFunction` is just the extract final results part. Since you need to do special handling on v2 side, is it possible to directly use the `ArrayAgg`? If not, probably extend that and only override the extract final results

**Comment by xiangfu0** [comment]
> The difference for `ListAggFunction` and `ArrayAggString`/`ArrayAggDistinctStringFunction` is just the extract final results part. Since you need to do special handling on v2 side, is it possible to directly use the `ArrayAgg`? If not, probably extend that and only override the extract final results

`ArrayAgg` doesn't support `ORDER BY` so the merger/extractFinal at intermediate/final stage doesn't honer the ordering.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java (hunk 6 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java]: Moved to another class to handle distinct

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 9]
```diff
@@ -47,11 +49,16 @@ public class AggregationFunctionFactory {
...
+  public static AggregationFunction getAggregationFunction(FunctionContext function, boolean nullHandlingEnabled) {
+    return getAggregationFunction(function, false, nullHandlingEnabled);
+  }
+
   /**
    * Given the function information, returns a new instance of the corresponding aggregation function.
    * <p>NOTE: Underscores in the function name are ignored in V1.
    */
-  public static AggregationFunction getAggregationFunction(FunctionContext function, boolean nullHandlingEnabled) {
+  public static AggregationFunction getAggregationFunction(FunctionContext function, boolean isDistinct,
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:60]: Let's not change this signature, and use argument to pass distinct similar to `ARRAYAGG`. In the future, we don't want to handle distinct logic within the aggregation function

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public T set(final int index, final T element) {
+      throw new UnsupportedOperationException("set element with index operation is not supported");
+    }
+
+    public boolean contains(Object element) {
+      return _hashSet.contains(element);
+    }
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java:88]: Are they used?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+  /**
+   * Custom implementation of {@link ObjectArrayList} that maintains an extra {@link ObjectOpenHashSet} to ensure that
+   * the elements are unique.
+   * Note that this implementation only overrides the methods used in {@link ListAggFunction}, so the uniqueness check
+   * is not guaranteed, use with your own risk.
+   *
+   * @param <T> Type of the element
+   */
+  static class ObjectArrayListHashSet<T> extends ObjectArrayList<T> {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java:63]: This is essentially an `ObjectLinkedOpenHashSet`. Can we use it as the intermediate result instead?

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  @Override
+  public ObjectArrayList<String> merge(ObjectArrayList<String> intermediateResult1,
+      ObjectArrayList<String> intermediateResult2) {
+    if (intermediateResult1 == null) {
+      return intermediateResult2;
+    }
+    if (intermediateResult2 == null) {
+      return intermediateResult1;
+    }
+    intermediateResult1.addAll(intermediateResult2);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java:137]: Will the intermediate result ever transferred to another node? If so, this will break because distinct is not honored

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java (hunk 10 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java]: Removed.

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java (hunk 11 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java]: Replaced ObjectArrayListHashSet with ObjectLinkedOpenHashSet

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java (hunk 9 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java]: This is the error for the query:
```
SELECT 
    Origin, Dest,
    LISTAGG(Carrier, ', ', 'true') WITHIN GROUP (ORDER BY Carrier)
FROM 
    airlineStats
GROUP BY 
    Dest, Origin
```
```
ProcessingException(errorCode:150, message:SQLParsingError:
java.lang.Exception: Unable to find table for this query
	at org.apache.pinot.controller.api.resources.PinotQueryResource.getMultiStageQueryResponse(PinotQueryResource.java:222)
	at org.apache.pinot.controller.api.resources.PinotQueryResource.executeSqlQuery(PinotQueryResource.java:179)
	at org.apache.pinot.controller.api.resources.PinotQueryResource.handlePostSql(PinotQueryResource.java:127)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
...
Caused by: java.lang.RuntimeException: Error composing query plan for: SELECT 
    Origin, Dest,
    LISTAGG(Carrier, ', ', 'true') WITHIN GROUP (ORDER BY Carrier)
FROM 
    airlineStats
...
Caused by: org.apache.calcite.runtime.CalciteContextException: From line 3, column 5 to line 3, column 34: Invalid number of arguments to function 'LISTAGG'. Was expecting 1 arguments
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
...
Caused by: org.apache.calcite.sql.validate.SqlValidatorException: Invalid number of arguments to function 'LISTAGG'. Was expecting 1 arguments
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490))
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 13]
```diff
@@ -254,15 +274,15 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
                 "ARRAY_AGG expects the 2nd argument to be literal, got: %s. The function can be used as "
                     + "arrayAgg(dataColumn, 'dataType', ['isDistinct'])", dataTypeExp.getType());
             DataType dataType = DataType.valueOf(dataTypeExp.getLiteral().getStringValue().toUpperCase());
-            boolean isDistinct = false;
+            boolean isDistinctArrayAgg = false;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:277]: (minor) You may revert this change if you put braces around the `LISTAGG` case. I feel it is easier to read to keep these variables local to the case

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 14]
```diff
@@ -245,6 +247,24 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
                 throw new IllegalArgumentException("Unsupported data type for FIRST_WITH_TIME: " + dataType);
             }
           }
+          case LISTAGG:
+            Preconditions.checkArgument(numArguments == 2 || numArguments == 3,
+                "LISTAGG expects 2 arguments, got: %s. The function can be used as "
+                    + "listAgg(['distinct'] expression, 'separator')", numArguments);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:253]:
```suggestion
                    + "listAgg([distinct] expression, 'separator')", numArguments);
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java [hunk 15]
```diff
@@ -245,6 +247,24 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
...
+            ExpressionContext separatorExpression = arguments.get(1);
+            Preconditions.checkArgument(separatorExpression.getType() == ExpressionContext.Type.LITERAL,
+                "LISTAGG expects the 2nd argument to be literal, got: %s. The function can be used as "
+                    + "listAgg(['distinct'] expression, 'separator')", separatorExpression.getType());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:257]:
```suggestion
                    + "listAgg([distinct] expression, 'separator')", separatorExpression.getType());
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+ */
+public class ListAggDistinctFunction extends ListAggFunction {
+
+  public ListAggDistinctFunction(ExpressionContext expression, String separator, boolean nullHandlingEnabled) {
+    super(expression, separator, nullHandlingEnabled);
+  }
+
+  @Override
+  protected AbstractObjectCollection<String> getObjectCollection(AggregationResultHolder aggregationResultHolder) {
+    ObjectLinkedOpenHashSet<String> valueSet = aggregationResultHolder.getResult();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java:40]: Currently `ObjectSerDeUtils` will ser/de this the same way as `ObjectOpenHashSet`, and lose the ordering. We need to separate them in `ObjectSerDeUtils`.
It would be good if we can make a test that fails when the intermediate result is serialized as `ObjectOpenHashSet`

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 17]
```diff
@@ -296,11 +299,25 @@ private TransferableBlock consumeAggregation() {
...
 
+  private static void handleListAggDistinctArg(String functionName, RexExpression.FunctionCall functionCall,
+      List<ExpressionContext> arguments) {
+    String upperCaseFunctionName =
+        AggregationFunctionType.getNormalizedAggregationFunctionName(functionName);
+    if (upperCaseFunctionName.equals("LISTAGG")) {
+      if (functionCall.isDistinct()) {
+        arguments.add(ExpressionContext.forLiteralContext(Literal.boolValue(true)));
+      } else {
+        arguments.add(ExpressionContext.forLiteralContext(Literal.boolValue(false)));
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:316]: (minor) Do we need to add a third argument?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java (hunk 16 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java:40]: Added a new type: `ORDERED_STRING_SET`
